File: ml-service/app/feature_schema.py
"""Canonical Feature Schema — Python mirror of server/src/schemas/featureSchema.ts
(TALWEG Final Upgrade Spec §4.1).

Shared conceptually by ingestion, ML, API and UI. Every value carries
provenance and quality status. `null` means missing — NEVER zero-filled.
"""

from __future__ import annotations

from typing import Literal, Optional, Union

from pydantic import BaseModel, Field

FEATURE_SCHEMA_VERSION = "1.0.0"

ProvenanceType = Literal["REAL", "DERIVED", "SYNTHETIC", "UNKNOWN"]
QualityStatus = Literal["VALID", "PARTIAL", "MISSING", "INVALID"]
AlignmentStatus = Literal["aligned", "partial", "unknown", "invalid"]

# Core prototype inputs used for completeness accounting (spec §4.1)
CORE_FEATURE_KEYS = (
    "rainfall_24h",
    "rainfall_3d",
    "slope",
    "soil_moisture",
    "historical_density",
)


class FeatureValue(BaseModel):
    value: Optional[Union[float, int, str]] = None
    unit: str
    source_id: str
    provenance_type: ProvenanceType = "UNKNOWN"
    observed_at: Optional[str] = None
    window_start: Optional[str] = None
    window_end: Optional[str] = None
    spatial_reference: Optional[str] = None
    spatial_distance_km: Optional[float] = None
    transformation: Optional[str] = None
    quality_status: QualityStatus = "VALID"


class FeatureLocation(BaseModel):
    latitude: float = Field(..., ge=-90.0, le=90.0)
    longitude: float = Field(..., ge=-180.0, le=180.0)
    zone_id: Optional[str] = None


class FeatureRecord(BaseModel):
    feature_schema_version: str = FEATURE_SCHEMA_VERSION
    location: FeatureLocation
    as_of: str
    features: dict[str, Optional[FeatureValue]] = Field(default_factory=dict)
    completeness: float = Field(0.0, ge=0.0, le=1.0)
    alignment_status: AlignmentStatus = "unknown"


def make_feature_value(
    value: Optional[Union[float, int, str]],
    *,
    unit: str,
    source_id: str,
    provenance: ProvenanceType,
    transformation: Optional[str] = None,
    observed_at: Optional[str] = None,
    window_start: Optional[str] = None,
    window_end: Optional[str] = None,
) -> FeatureValue:
    """Factory mirroring makeFeatureValue(): null → MISSING; non-finite → INVALID."""
    if value is None:
        quality: QualityStatus = "MISSING"
    elif isinstance(value, float) and (value != value or value in (float("inf"), float("-inf"))):
        quality = "INVALID"
    else:
        quality = "VALID"
    return FeatureValue(
        value=value,
        unit=unit,
        source_id=source_id,
        provenance_type=provenance,
        observed_at=observed_at,
        window_start=window_start,
        window_end=window_end,
        transformation=transformation,
        quality_status=quality,
    )


def compute_completeness(features: dict[str, Optional[FeatureValue]]) -> float:
    """Fraction of core features that are present and VALID."""
    valid = 0
    for key in CORE_FEATURE_KEYS:
        fv = features.get(key)
        if fv is not None and fv.value is not None and fv.quality_status == "VALID":
            valid += 1
    return round(valid / len(CORE_FEATURE_KEYS), 3)


def compute_alignment_status(features: dict[str, Optional[FeatureValue]]) -> AlignmentStatus:
    saw_partial = False
    for key in CORE_FEATURE_KEYS:
        fv = features.get(key)
        if fv is None or fv.value is None:
            continue
        if fv.quality_status == "INVALID":
            return "invalid"
        if fv.provenance_type == "UNKNOWN":
            return "unknown"
        if fv.provenance_type in ("DERIVED", "SYNTHETIC"):
            saw_partial = True
        if fv.spatial_distance_km is not None and fv.spatial_distance_km > 0:
            saw_partial = True
    return "partial" if saw_partial else "aligned"

File: server/src/schemas/featureSchema.ts
/**
 * Canonical Feature Contract (Final Upgrade Spec §4.1)
 *
 * One schema shared conceptually by ingestion, ML, API and UI.
 * Every environmental value carries its own provenance, quality status and
 * (where applicable) temporal/spatial alignment metadata.
 *
 * HARD RULES (spec §4.1):
 * - Do NOT force unavailable fields to zero. `0` means an observed zero;
 *   `null` means missing.
 * - Every value must be traceable to a source (`source_id`) and a
 *   provenance type (REAL | DERIVED | SYNTHETIC | UNKNOWN).
 */

export const FEATURE_SCHEMA_VERSION = '1.0.0';

export type ProvenanceType = 'REAL' | 'DERIVED' | 'SYNTHETIC' | 'UNKNOWN';

export type QualityStatus = 'VALID' | 'PARTIAL' | 'MISSING' | 'INVALID';

export type AlignmentStatus = 'aligned' | 'partial' | 'unknown' | 'invalid';

export interface FeatureValue {
  /** null = missing (never substitute 0 for a missing observation) */
  value: number | string | null;
  unit: string;
  source_id: string;
  provenance_type: ProvenanceType;
  observed_at?: string;
  window_start?: string;
  window_end?: string;
  /** e.g. 'EPSG:4326', 'SRTM 30m DEM' */
  spatial_reference?: string;
  /** distance between observation location and target location, km */
  spatial_distance_km?: number;
  /** transformation applied (e.g. 'rolling_sum_3d', 'exp_decay_tau3d') */
  transformation?: string;
  quality_status: QualityStatus;
}

export interface FeatureRecordLocation {
  latitude: number;
  longitude: number;
  zone_id?: string;
}

/**
 * Canonical feature record. All fields optional — absent means "not
 * available for this location/time", never zero.
 */
export interface FeatureRecord {
  feature_schema_version: string;
  location: FeatureRecordLocation;
  as_of: string;
  features: {
    // Rainfall family
    rainfall_24h?: FeatureValue;
    rainfall_3d?: FeatureValue;
    rainfall_5d?: FeatureValue;
    rainfall_7d?: FeatureValue;
    rainfall_intensity?: FeatureValue;
    rainfall_percentile?: FeatureValue;
    rainfall_anomaly?: FeatureValue;
    antecedent_rainfall_index?: FeatureValue;
    threshold_exceedance_ratio?: FeatureValue;
    threshold_exceedance_duration?: FeatureValue;
    // Terrain family
    slope?: FeatureValue;
    aspect?: FeatureValue;
    elevation?: FeatureValue;
    curvature?: FeatureValue;
    terrain_ruggedness?: FeatureValue;
    relative_relief?: FeatureValue;
    // Soil family
    soil_moisture?: FeatureValue;
    soil_type?: FeatureValue;
    soil_depth?: FeatureValue;
    sand_fraction?: FeatureValue;
    silt_fraction?: FeatureValue;
    clay_fraction?: FeatureValue;
    bulk_density?: FeatureValue;
    hydraulic_conductivity?: FeatureValue;
    available_water_capacity?: FeatureValue;
    organic_carbon?: FeatureValue;
    // History / context family
    historical_density?: FeatureValue;
    distance_to_historical_event?: FeatureValue;
    geology_class?: FeatureValue;
    land_cover?: FeatureValue;
    vegetation_index?: FeatureValue;
  };
  /** fraction [0,1] of expected core features that are VALID */
  completeness: number;
  alignment_status: AlignmentStatus;
}

/** Keys considered core for risk computation completeness (5 prototype inputs + rainfall memory). */
export const CORE_FEATURE_KEYS = [
  'rainfall_24h',
  'rainfall_3d',
  'slope',
  'soil_moisture',
  'historical_density',
] as const;

/** All canonical feature keys, for completeness/quality accounting. */
export const ALL_FEATURE_KEYS = [
  'rainfall_24h', 'rainfall_3d', 'rainfall_5d', 'rainfall_7d',
  'rainfall_intensity', 'rainfall_percentile', 'rainfall_anomaly',
  'antecedent_rainfall_index', 'threshold_exceedance_ratio', 'threshold_exceedance_duration',
  'slope', 'aspect', 'elevation', 'curvature', 'terrain_ruggedness', 'relative_relief',
  'soil_moisture', 'soil_type', 'soil_depth', 'sand_fraction', 'silt_fraction', 'clay_fraction',
  'bulk_density', 'hydraulic_conductivity', 'available_water_capacity', 'organic_carbon',
  'historical_density', 'distance_to_historical_event', 'geology_class', 'land_cover', 'vegetation_index',
] as const;

export type FeatureKey = (typeof ALL_FEATURE_KEYS)[number];

/**
 * Convenience factory for a numeric feature value.
 * `null` value → quality_status MISSING (0 is never substituted).
 */
export function makeFeatureValue(
  _key: string,
  value: number | string | null,
  opts: {
    unit: string;
    sourceId: string;
    provenance: ProvenanceType;
    observedAt?: string;
    windowStart?: string;
    windowEnd?: string;
    spatialReference?: string;
    spatialDistanceKm?: number;
    transformation?: string;
    qualityStatus?: QualityStatus;
  }
): FeatureValue {
  let quality = opts.qualityStatus ?? 'VALID';
  if (value === null || value === undefined) {
    quality = 'MISSING';
  } else if (typeof value === 'number' && !Number.isFinite(value)) {
    quality = 'INVALID';
  }
  const fv: FeatureValue = {
    value: value ?? null,
    unit: opts.unit,
    source_id: opts.sourceId,
    provenance_type: opts.provenance,
    quality_status: quality,
  };
  if (opts.observedAt) fv.observed_at = opts.observedAt;
  if (opts.windowStart) fv.window_start = opts.windowStart;
  if (opts.windowEnd) fv.window_end = opts.windowEnd;
  if (opts.spatialReference) fv.spatial_reference = opts.spatialReference;
  if (opts.spatialDistanceKm !== undefined) fv.spatial_distance_km = opts.spatialDistanceKm;
  if (opts.transformation) fv.transformation = opts.transformation;
  return fv;
}

/**
 * Completeness = fraction of CORE features that are present and VALID.
 * Computed only over the 5 prototype core inputs.
 */
export function computeCompleteness(features: FeatureRecord['features']): number {
  let valid = 0;
  for (const key of CORE_FEATURE_KEYS) {
    const fv = features[key];
    if (fv && fv.value !== null && fv.quality_status === 'VALID') valid += 1;
  }
  return Math.round((valid / CORE_FEATURE_KEYS.length) * 1000) / 1000;
}

/**
 * Alignment status derived from the underlying value alignments:
 * - 'aligned'  : all present core values observed at/for the requested location & time
 * - 'partial'  : some values approximated (spatially or temporally) or derived
 * - 'unknown'  : provenance unknown for any core value
 * - 'invalid'  : any core value INVALID
 */
export function computeAlignmentStatus(features: FeatureRecord['features']): AlignmentStatus {
  let sawPartial = false;
  for (const key of CORE_FEATURE_KEYS) {
    const fv = features[key];
    if (!fv || fv.value === null) continue; // missing handled by completeness
    if (fv.quality_status === 'INVALID') return 'invalid';
    if (fv.provenance_type === 'UNKNOWN') return 'unknown';
    if (fv.provenance_type === 'DERIVED' || fv.provenance_type === 'SYNTHETIC') sawPartial = true;
    if (fv.spatial_distance_km !== undefined && fv.spatial_distance_km > 0) sawPartial = true;
  }
  return sawPartial ? 'partial' : 'aligned';
}
